src/darwin/Framework/CHIP/MTRDevice.h

```diff
@@ -240,6 +240,90 @@ typedef NS_ENUM(NSUInteger, MTRDeviceState) {
                                       completion:(MTRDeviceOpenCommissioningWindowHandler)completion
     MTR_AVAILABLE(ios(17.0), macos(14.0), watchos(10.0), tvos(17.0));
 
+/**
+ *
+ * This set of functions allows clients to store metadata for either an entire device or for a specific endpoint.
+ *
+ * Notes:
+ *   • Client data will be removed automatically when devices are deleted from the fabric
+ *   • Supported client data object types are currently only:
+ *         NSData, NSString, NSArray, NSDictionary, NSNumber
+ */
+
+/**
+ *
+ * List of all client data types supported
+ *
+ */
+- (NSArray *)supportedClientDataClasses MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * List of all client data keys stored
+ *
+ */
+- (NSArray * _Nullable)clientDataKeys MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * Retrieve client metadata for a key, returns nil if no value is set
+ *
+ * @param key           NSString * for the key to store the value as
+ */
+- (id<NSSecureCoding> _Nullable)clientDataForKey:(NSString *)key MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * Set client metadata for a key. The value must conform to NSSecureCoding
+ *
+ * @param key           NSString * for the key to store the value as
+ * @param value         id <NSSecureCoding> for the value to store
+ */
+- (void)setClientDataForKey:(NSString *)key value:(id<NSSecureCoding>)value MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * Remove client metadata for a key.
+ *
+ * @param key           NSString * for the key to store the value as
+ */
+- (void)removeClientDataForKey:(NSString *)key MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * List of all client data keys stored
+ *
+ */
+- (NSArray * _Nullable)clientDataKeysForEndpointID:(NSNumber *)endpointID MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * Retrieve client metadata for a key, returns nil if no value is set
+ *
+ * @param key           NSString * for the key to store the value as
+ * @param endpointID    NSNumber * for the endpoint to associate the metadata with
+ */
+- (id<NSSecureCoding> _Nullable)clientDataForKey:(NSString *)key endpointID:(NSNumber *)endpointID MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * Set client metadata for a key. The value must conform to NSSecureCoding.
```

Comment:
https://github.com/project-chip/connectedhomeip/pull/30730#discussion_r1410850766 does not seem to have been addresse?